### ISSUE TITLE
Dictionaries not working as expected

### ISSUE THREAD

**Issue by joni1993** [Icinga/icingaweb2-module-director#2275]
# Issue 1 - Can't add entry to dictionary
## Expected Behavior
After defining a Host template and assigning fields that are later used for the dictionary (as template), i assign this dictionary field to another Host. On the actual host object then add fields to this dictionary as they were specified in the template.

## Current Behavior
An error occurs when trying to add a field to the host objects dictionary.

## Possible Solution
n/a - see disclaimer

## Steps to Reproduce (for bugs)
- add data field "address_ip" of type string
- add data field "address_type" of type string
- add host template "Additional IP Address Host" with previously created fields
- add data field "additional_addresses" of type dictionary with template  "Additional IP Address Host"
- create host template "Router" with new dictionary data field "additional_addresses"
- create host that imports "Router"
- click "Manage Instances" and then "add"
- following error appears: 
```
Unable to load parent referenced from service "", failed to load icinga_service for host_id IS NULL AND service_set_id IS NULL AND object_name = 'Additional IP Address Host' AND object_type = 'template' (IcingaObjectImports.php:266)
```

# Issue 2 - Error if template used in dictionary is renamed
## Steps to Reproduce
- same as Issue 1
- rename "Additional IP Address Host" to "additional-ip-host"
- when you click "Manage Instances" now, the following exception appears
```


Failed to load icinga_host "{"object_name":"Additional IP Address Host"}"

#0 /usr/share/icingaweb2/modules/director/library/Director/Data/Db/DbObject.php(1173): Icinga\Module\Director\Data\Db\DbObject->loadFromDb()
#1 /usr/share/icingaweb2/modules/director/application/controllers/DataController.php(193): Icinga\Module\Director\Data\Db\DbObject::load()
#2 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Director\Controllers\DataController->dictionaryAction()
#3 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch()
#4 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch()
#5 /usr/share/php/Icinga/Application/Web.php(300): Zend_Controller_Front->dispatch()
#6 /usr/share/php/Icinga/Application/webrouter.php(99): Icinga\Application\Web->dispatch()
#7 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#8 {main}
```
## Workarround
Just set the template again in the dictionary data field

## Solution
Do not allow renaming a template when used by a dictionary field

# Issue 3 - Wrong object loaded if dictionary template is of type "services" but dictionary is assigned to a "host template"
## Steps to Reproduce
- same as Issue 1
- change type of dictionary field  "additional_addresses" to service
- use any service template (i used a service template for Ping4 service) that has fields assigned
- go to the host that uses this field and click "Manage Instances"
- following error appears:
```
Failed to load icinga_host "{"object_name":"Ping4"}"

#0 /usr/share/icingaweb2/modules/director/library/Director/Data/Db/DbObject.php(1173): Icinga\Module\Director\Data\Db\DbObject->loadFromDb()
#1 /usr/share/icingaweb2/modules/director/application/controllers/DataController.php(193): Icinga\Module\Director\Data\Db\DbObject::load()
#2 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Director\Controllers\DataController->dictionaryAction()
#3 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch()
#4 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch()
#5 /usr/share/php/Icinga/Application/Web.php(300): Zend_Controller_Front->dispatch()
#6 /usr/share/php/Icinga/Application/webrouter.php(99): Icinga\Application\Web->dispatch()
#7 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#8 {main}
```

## Your Environment

* Director version (System - About):
master (e2a88861487e9cefad94776fd610b3a9703b9e03)
* Icinga Web 2 version and modules (System - About):
2.8.2
* Icinga 2 version (`icinga2 --version`):
r2.12.3-1
* Operating System and version:
Ubuntu 18.04.5
* Webserver, PHP versions:
Apache + PHP 7.4.13
Related:
#337 
40e1b5a798b8d2a2853494c220ece009af2b32ca

Disclaimer: I am not sure if I understood everything of this new feature completely. I suspect the "Template (Object) Type"+Template are used for the general "inside" structure of the dictionary (based on the fields in this template) - If this is correct why are "Host/Service Templates" used for this, as a dictionary normally has nothing to do directly with a Host/Service template? Or is it just that not another "object type" like "dictionary template" has to be implemented - wouldn't that be a finer solution, as my "Additional IP Address Host" is not used for anything but serving as a holder for fields of the dictionary?

**Comment by drallgood** [comment]
Actually it looks like you can't use it for anything but services/hosts. I tried using it for a notification template and it fails because "Required parameter 'host' missing"

**Comment by joni1993** [comment]
I created a gist for the first issue - this allows the usage of dictionaries for hosts. https://gist.github.com/joni1993/df1566eb1c2a8db425e55f3c6b7ff07f

I did not create PR because i do not know if given behaviour is indented - but as you have a switcher when creating the dictionary field (enum=>service or host), i think it should work for hosts too.

Inspecting the code it seemed like it was hard coded to only work for services (IcingaServiceDictionaryMemberForm.php) - no IcingaHostDictionaryMemberForm.php here. So i generalized that, to work for hosts and services.

But as long as you cant choose the dictionary type in the apply-for section, it is not very useful at all - you cant create a service-apply for those fields.
